pose_prediction/const_vel/plugin.cpp

```diff
@@ -0,0 +1,410 @@
+    }
+
+    virtual Eigen::Quaternionf get_offset() override {
+        return offset;
+    }
+
+    // Correct the orientation of the pose due to the lopsided IMU in the
+    // current Dataset we are using (EuRoC)
+    virtual pose_type correct_pose(const pose_type pose) const override {
+        pose_type swapped_pose;
+
+        // Make any changes to the axes direction below
+        // This is a mapping between the coordinate system of the current
+        // SLAM (OpenVINS) we are using and the OpenGL system.
+        swapped_pose.position.x() = -pose.position.y();
+        swapped_pose.position.y() = pose.position.z();
+        swapped_pose.position.z() = -pose.position.x();
+
+        // Make any chanes to orientation of the output below
+        // For the dataset were currently using (EuRoC), the output orientation acts as though
+        // the "top of the head" is the forward direction, and the "eye direction" is the up direction.
+        Eigen::Quaternionf raw_o (pose.orientation.w(), -pose.orientation.y(), pose.orientation.z(), -pose.orientation.x());
+
+        swapped_pose.orientation = apply_offset(raw_o);
+        swapped_pose.sensor_time = pose.sensor_time;
+
+        return swapped_pose;
+    }
+
+private:
+    mutable std::atomic<bool> first_time{true};
+    const std::shared_ptr<switchboard> sb;
+    const std::shared_ptr<const RelativeClock> _m_clock;
+    switchboard::reader<pose_type> _m_slow_pose;
+    switchboard::reader<imu_raw_type> _m_imu_raw;
+    switchboard::reader<pose_type> _m_true_pose;
+    switchboard::reader<switchboard::event_wrapper<Eigen::Vector3f>> _m_ground_truth_offset;
+    switchboard::reader<switchboard::event_wrapper<time_point>> _m_vsync_estimate;
+    mutable Eigen::Quaternionf offset {Eigen::Quaternionf::Identity()};
+    mutable std::shared_mutex offset_mutex;
+
+    // Based on https://github.com/rpng/open_vins/issues/92
+    // Returns a pair of the predictor state and the time associated with the
+    // most recent imu reading used to perform this prediction.
+    std::pair<Eigen::Matrix<double, 13, 1>, time_point> predict_const_vel(double dt) const {
+        const auto imu_raw = _m_imu_raw.get_ro();
+
+        // Current state
+        const auto curr_time = imu_raw->imu_time;
+        // TODO: should this be the last w or the estimated w in the VIO/integrator state?
+        const auto curr_w    = 0.5 * (imu_raw->w_hat + imu_raw->w_hat2);
+        const auto curr_quat = imu_raw->quat;
+        const auto curr_R    = quat_2_Rot(Eigen::Matrix<double, 4, 1>{curr_quat.x(), curr_quat.y(), curr_quat.z(), curr_quat.w()});
+        const auto curr_pos  = imu_raw->pos;
+        const auto curr_vel  = imu_raw->vel;
+
+        // Compute new quaternion
+
+        // Option 1: PG's code from issue 92. Same as Propagator.cpp
+        const auto future_q = rot_2_quat(exp_so3(-curr_w * dt) * curr_R);
+
+        // Option 2: PG's formula from issue 92. Same as Trawny eq. 96.
```

Comment:
What's the trade-offs between these two options?
Which one should be picked eventually?